### ISSUE TITLE
RFC: Deprecate support for LLVM <= 5

### ISSUE THREAD

**Issue by elliottslaughter** [terralang/terra#471]
**Update:** we are in phase 4 below. LLVM <= 5 has been formally deprecated and is planned for removal in late 2022. Please comment on this issue if you are still using LLVM 3.8 in particular (as this is the last release to support NVVM, prior to our switch to the NVPTX backend with LLVM 5+). Also, if you have not had an opportunity to directly compare LLVM 3.8 and newer versions, please take the time to do so.

I'm going to kickstart this discussion early since I know there are still users on LLVM 3.8. I would ideally like to move our minimum LLVM version to 6, i.e. deprecating all versions <= 5.

This will be similar to https://github.com/terralang/terra/issues/456, except that (since I know there are live users on 3.8) I want to make sure we confirm all those users have migrated before we pull the plug.

Steps:

 1. \[DONE] Verify all users have moved to at least LLVM 6 and there are no outstanding issues associated with those newer versions.
 2. \[DONE] Wait a couple of months to be sure we're not missing anything.
 3. \[DONE] Formally deprecate LLVM <= 5. This means changing the README and release notes. No code changes.
 4. \[DONE] Wait an additional 6 months before removing the code from the repository.

If you are actively using any LLVM versions <= 5, please reply to this issue as soon as possible so that your usage can be tracked.

Blockers:

- [x] #467
- [x] #472
- [x] https://github.com/StanfordLegion/legion/issues/983
- [x] https://github.com/StanfordLegion/legion/issues/1041
- [x] #527
- [x] #530

**Comment by mariodirenzo** [comment]
We have issues with LLVM 6 on Yellowstone at Stanford. However LLVM 3.8 works just fine.
Maybe @cmelone can comment on this

**Comment by cmelone** [comment]
The solver builds successfully on LLVM 6, but when we run test cases, this error occurs:

```
Invoking Legion on 1 rank(s), 1 node(s) (1 rank(s) per node), as follows:
/home/hpcc/cmelone/insieme/builds/890523826/HTR-solver/src/prometeo_ConstPropMix.exec -i TaylorGreen2D_16.json -o 16 -ll:cpu 1 -ll:ocpu 2 -ll:onuma 1 -ll:othr 9 -ll:ostack 8 -ll:gpu 1 -ll:fsize 4000 -ll:zsize 512 -ll:ib_zsize 512 -ll:util 2 -ll:io 1 -ll:bgwork 2 -ll:cpu_bgwork 100 -ll:util_bgwork 100 -ll:csize 220000 -ll:rsize 512 -ll:ib_rsize 512 -ll:gsize 0 -ll:stacksize 8 -lg:sched -1 -lg:hysteresis 0
[0 - 7f0af0f41840]    0.744143 {5}{gpu}: Failed to load CUDA module! Error log: ptxas application ptx input, line 29155; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 29279; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 29440; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 29709; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30067; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30129; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30206; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30270; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30336; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30408; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 30502; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 36620; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 36753; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 37325; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas application ptx input, line 37604; error   : Feature 'rsqrt.approx.ftz.f64' requires PTX ISA .version 4.0 or later
ptxas fatal   : Ptx assembly aborted due to errors
```

I'm not sure how related this is directly to terra, but this is an issue we encounter on 6.0 and not 3.8

**Comment by elliottslaughter** [comment]
@cmelone @mariodirenzo Thanks for reporting. I'm splitting this off into a separate issue so that we can track it. I'll mark it as a blocker up above.

**Comment by elliottslaughter** [comment]
There is one large caveat for CUDA users: Terra switched to NVPTX starting with LLVM 5, and that means that certain details of the backends, like what intrinsics are available, differ. See https://github.com/StanfordLegion/legion/issues/983 an example of the sort of issues you might face.

I think think means we'll want to be extra sure we compare performance and compatibility with users of CUDA in Terra.

**Comment by elliottslaughter** [comment]
CC'ing more people who might have a stake in the CUDA backend: @jameshegarty @gilbo @Mx7f @ProfFan @aiverson @ErikMcClure 

Please see above discussion, particularly about the migration from NVVM to NVPTX, and let us know (a) if this impacts you, and (b) what if any testing you would like to do before LLVM 3.8 (and therefore NVVM support) is deprecated.

**Comment by elliottslaughter** [comment]
Now that I've replicated LLVM 3.8 performance on the apps I've been testing, I am formally kicking off the deprecation period for LLVM <= 5. This puts us in phase 1 above. I have a couple of known users on LLVM 3.8 (via Regent) who I will contact. **If anyone else is still using any LLVM version less than or equal to 5, speak up now.** This phase will be relatively short (assuming the known users confirm they can move to newer LLVM versions), and then we'll be on to phase 2.

**Note that if you're upgrading, there are some things that have changed recently that will (hopefully) improve your performance, especially on NVIDIA GPUs.** Some of these things are automatic, but some may require changes on your end. Specifically:

 * The LLVM inliner pass was previously disabled for NVPTX with LLVM 5 and newer. This has been fixed. Note that this may or may not impact performance as PTX has its own inliner.
 * The atomic intrinsics in NVVM were never supported by NVPTX (and therefore, Terra with LLVM 5 or newer). Users were either forced to stick with LLVM 3.8 or switch to inline assembly. You would know if this is you (assuming you've attempted the upgrade). Terra now directly supports the `atomicrmw` instruction for use with newer LLVM versions. (See [here](https://terralang.org/api.html#built-in-macros) and scroll down to `atomicrmw`.)
 * NVVM used the equivalent of LLVM's `contract` fast-math option by default. NVPTX does not. Terra now has explicit support for passing an "optimization profile" to `terralib.saveobj` and `cudalib.toptx`, which can include fast-math flags. To achieve the equivalent of the old NVVM performance, pass `{fastmath="contract"}`. You might see additional performance gains by passing `{fastmath=true}` to enable all optimizations. Note that in general fast-math optimizations can introduce numerical inaccuracy and may not be safe with every application.

I'm CCing a list of people who have been interested in CUDA support at some point in the past. If you are still using LLVM 3.8, or have not done a direct comparison to newer versions to verify the performance, now is the time to do so.  @jameshegarty @gilbo @Mx7f @ProfFan @aiverson @ErikMcClure

**Comment by elliottslaughter** [comment]
I haven't heard anything, and all the Regent users I've talked to have had performance that is at least as good (and other issues have been resolved), so I'm going to kick off the next phase.

I might accelerate this a bit so we get the deprecation in before 1.0.0, but otherwise things will be as before: we'll wait a bit longer to see if there's any reason not to deprecate, then formalize the deprecation. Nothing will be removed at this point (and not before 1.0.0 at any rate).
